### ISSUE TITLE
Clip masks incorrectly applied within nested groups

### ISSUE THREAD

**Issue by davidyxu** [paperjs/paper.js#1594]
# Description/Steps to reproduce
Clip masks are being incorrectly applied when inside groups with applyMatrix set to false and `opacity !== 1`.

# Link to reproduction test-case
http://sketch.paperjs.org/#V/0.11.8/S/jVSxbtswEP2VgxZLgCHLQRHAMTxlyJICRTN0iDLQ0tlizRwJkopqBPn3HinLlRy3qAZDvne89/juQe8JiVdM7pKnA/qqSeZJpevw3wiDNnfovaS9y4Ux6vhVeCt/wQZ2Qjlcl1TSYgGirkFApaR5Fe4AXoNvEETl5RuCEke0JVWanIfKasOnCTt4anh+/h0rL2ivMC0J+AnINy3Jp8Ucimx+Wb0tuP6lKLKSMqY3Vv/kCXnP9Rio8qDDYM003rZnja7RHYvcalujhZ22UeOgedB3gjdRKE/ShGng6eu581Yf8IesfcM9y0vgXisdDs8s1rMzMfrWMDOh86xqb3VrStq1xJI19fBDKKYZvPfX7bXo1jtZY8ROnp361uM2Sf/fFT3/p/9/38GVPdzGPQRgIOt/R2T5Tip19mVvEWl2pTdKzjlH941UdToaMEweu3HZGItZNDy0nk0PG3aVUJxf5uEddOIIXYMWJ8sH6VyL0LFS2CLo7ZvUrfssL4zC9Oaqoh5bLpbFAFuWYGnSxcjHJBW+08AgecnH+2i4+UmhNqKS/shzXKu8CxeQnFJreWNRuuFrDbmNR5ds8ThO6wl68xnt6/nAtIEiX61Wf4AqeGyRnouXyftkpzHr/NXYWhQHE8Lhkrvnl4/f

# Expected result
There should just be a square that's almost solid red showing.

# Additional information
macOS 10.12.6, chrome 70.0.3538.77

It looks like it's caused by some logic in the draw function for `paper.Item`, I've made some changes here which seem to fix the problem without any other changes (as far as I can tell):
https://github.com/davidyxu/paper.js/commit/a5f0df87f39d1c76029ef15919c8fddabffcc736

**Comment by sasensi** [comment]
Hi, thanks for the report !
This should be fixed with #1596.
Instead of the fix you proposed, I simply inverted the order of the actions which fixes the bug too:
- first draw the clip item
- then apply transformation

**Comment by davidyxu** [comment]
Cool, looks good to me.

**Comment by davidyxu** [comment]
@sasensi actually after taking a closer look, I think that your fix doesn't handle all the issues

sorry for the complicated example, but you can take a look at this example, i've setup a scene where you can comment and uncomment line 163 to see what it looks like:
http://sketch.paperjs.org/#V/0.11.8/S/5Vffb9owEP5XrLyUSuA6ISGk1Z72MGlat0l9BB7cxC1ugx05ph1F/d9nE0JK7PxgVadJuwewfXdf7i6f7dzWYXhFnEvn5pHIeOkMnZgneh5zlkvwkHMGPoHZnAEwd77hDRFzZ6hnW/2jF7X/3LlUo5ivMp5TSfYmWouzLN1cYynoL210h9OcHJTxkqaJIExrZsUiqAba4ieWywOalm017IAvTVYH7eytAgB3eDxHHfO6/Sia1E1GgY/erixqscQpza5x/qijkWJdDzUn9yvCZG4JtjY1owMgiHzou1PkHisWw5OR0IkIru8HMAhDL/xIqKb82gvOc5KU5a50r+Wwcjji3c0SZ+SfJZ43DqHnuuN6jUae78EwQlHUVhO5yfYbVpBYYnafHme6YyJ9IbbIm96O9d3UHyxwQtc2chsFaEO5o2n6madc9KisMT8C7k+HL4KvszY6VIegJLk0yvkutsDJeDKZnMoZq9fID2GIIoNQfoDgOESe31Z4nuGYyo0O0a1vM9tBXoixw23VLGRbX+hVuh4xNETSFU9jVKVj9dapui4bEU5Io3cyrSmVKOYpZkpjen8UeenUyGZT6mQ0xbygTrfofkrQDRKgdov67VaXjg8Am0P3YW3zaji+TXFRj9Khd6fdePKb0uNlvzeathvE8jjooXDqd4cFvWA69XrZoWjakUSb+rVZ2Zb83zgsPpiuXg+6ev83XTXBgs6gTrD6IKI2qRoUViiLrbFUczzuFmxfgMVg96cWF1dzNmeZ4A+KzxDHkj6RXRsMqWp2hfx68+P7QP/AXN147J7ebQa6aT4/3zkWXbT+JvRUG22DKW96qI2g6lcYGRS+Oy8oBWZ5iiUZzBSzwRCghVYXygM1FPiZIMmZ0lxcgLV67kp3kkAuaQ5SygjALFEzApQVOOxN8KwAQEJzdckTLHbOBfL+Y0/hKgpcVRV4ouQZvnC+Uhr3wr9yhs6tIPgx41Q1rs7lbPH6Gw==

here's a couple screenshots of the different outputs

on 0.11.8:
![image](https://user-images.githubusercontent.com/1980008/48156083-618ba380-e281-11e8-856f-a3a36b3d11ab.png)

with your fix (reordering transform and clip):
![image](https://user-images.githubusercontent.com/1980008/48156106-6fd9bf80-e281-11e8-9caa-503e0a7ebcd0.png)

with my earlier fix (removing clip):
![image](https://user-images.githubusercontent.com/1980008/48156128-8122cc00-e281-11e8-833c-be93e62f7783.png)

**Comment by davidyxu** [comment]
@sasensi Yeah I see that issue as well, I'm not sure if it's the same issue though.

If you use blendmodes `hue` vs `normal` instead or change opacity between `1` and `0.5`, you'll see that the version with clip removed will give the same results between the direct and not direct codepaths in draw whereas the other version will change.

I wonder if your example using the `add` blendMode is revealing a separate bug?

**Comment by sasensi** [comment]
@davidyxu I think changing `blendMode` or `opacity` are the same bug because they both cause item drawing to be done in a separated canvas and that's the core of this issue.

**Comment by davidyxu** [comment]
Right, blendMode and opacity should both cause the same bug, but I believe the `hue` example I mentioned should also trigger the bug like the `add` blendmode, since it's also considered a non native blendmode.

It looks like the example only shows the issue when used with certain custom blend mode (`add`, `subtract`, `average`, and `negation`), but other custom blend modes are working ok. Similarly, the opacity change are also fixed by the clip change. That's why I'm not sure if the `add` blend mode is exactly the same issue.

**Comment by lehni** [comment]
@sasensi what's the state of this bug? We're still looking for a solution, right?

**Comment by sasensi** [comment]
@lehni, yes, I did not find the solution for this one.
